"Permission denied" when same name occurs twice in Cargo.toml

This relates to https://github.com/dfinity/agent-rs/pull/162

Basically the `Cargo.toml` of agent-rs had the following:
```
members = [
  ...
  "ic-types/",
  ...
]

[patch.crates-io]
ic-types = { path = "./ic-types", version = "0.1", features = [ "serde" ] }
```

This leads to a "permission denied" failure when using `naersk` to compile it. It was essentially trying to copy `ic_types` sub-directory twice to a target directory, and the second copying failed because file permission was `0444`.

The above reported problem is reproducible with commit version ff6a6c73e1e0ffbf179f6a7a063fbb64e16e9214.

The latest master has a different problem leading to unresolved imports:

```
error[E0432]: unresolved import `ic_types::Principal`
 --> /sources/candid-0.6.21/src/types/reference.rs:4:5
  |
4 | use ic_types::Principal;
  |     ^^^^^^^^^^^^^^^^^^^ no `Principal` in the root
```

I've made a draft PR #172 to reproduce this with master (but can also be rebased to ff6a6c73e1e0ffbf179f6a7a063fbb64e16e9214 to reproduce the above reported problem).

--- a/lib.nix
+++ b/lib.nix
@@ -196,14 +196,6 @@ rec
             ${lib.optionalString (! isNull cargoconfig) "cp ${config} $out/.cargo/config"}
             cp ${cargolock'} $out/Cargo.lock
 
-            for p in $copySources; do
-              echo "Copying patched source $p to $out..."
-              # Create all the directories but $p itself, so `cp -R` does the
-              # right thing below
-              mkdir -p "$out/$(dirname "$p")"
-              cp --no-preserve=mode -R "$copySourcesFrom/$p" "$out/$p"
-            done
-
             for tuple in $cargotomlss; do
                 member="''${tuple%%:*}"
                 cargotoml="''${tuple##*:}"
@@ -225,6 +217,20 @@ rec
                 # script which is `./build.rs`.
                 echo 'fn main(){}' > build.rs
                 popd > /dev/null
+            done
+
+            # Copy all the "patched" sources which are used by dependencies.
+            # This needs to be done after the creation of the dummy to make
+            # sure the dummy source files do not tramp on the patch
+            # dependencies.
+            for p in $copySources; do
+              echo "Copying patched source $p to $out..."
+              mkdir -p "$out/$p"
+
+              chmod -R +w "$out/$p"
+              echo copying "$copySourcesFrom/$p"/ to "$out/$p"
+
+              cp --no-preserve=mode -R "$copySourcesFrom/$p"/* "$out/$p"
             done
           '';
 

--- a/nix/sources.json
+++ b/nix/sources.json
@@ -1,4 +1,16 @@
 {
+    "agent-rs": {
+        "branch": "paulliu/add-cargo-lock",
+        "description": "A collection of libraries and tools for building software around the Internet Computer, in Rust.",
+        "homepage": "https://sdk.dfinity.org/",
+        "owner": "ninegua",
+        "repo": "agent-rs",
+        "rev": "4a22e590516bc79ec3c75a320f7941e7762ea098",
+        "sha256": "0sacddc34nlfgldqghlwchgzjki177h5dsgpmdv70cm8hfy0sg7l",
+        "type": "tarball",
+        "url": "https://github.com/ninegua/agent-rs/archive/4a22e590516bc79ec3c75a320f7941e7762ea098.tar.gz",
+        "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
+    },
     "cargo": {
         "branch": "nm/canonicalize-path-args",
         "description": "The Rust package manager",

--- a/test.nix
+++ b/test.nix
@@ -210,6 +210,18 @@ let
     talent-plan-3 = naersk.buildPackage "${sources.talent-plan}/rust/projects/project-3";
 
     rustlings = naersk.buildPackage sources.rustlings;
+
+    agent-rs = naersk.buildPackage {
+      doCheck = true;
+      src = sources.agent-rs;
+      buildInputs =
+        [
+          pkgs.openssl
+          pkgs.pkg-config
+          pkgs.perl
+        ] ++ pkgs.lib.optional pkgs.stdenv.isDarwin pkgs.libiconv;
+    };
+
   };
 in
 fastTests

